source/simulator/assembly.cc

```diff
@@ -613,26 +536,43 @@ namespace aspect
                      const double                        average_field,
                      const double                        global_entropy_variation,
                      const double                        cell_diameter,
-                     const AdvectionField     &advection_field) const
+                     const AdvectionField               &advection_field) const
   {
-    //discontinuous Galerkin doesn't require an artificial viscosity
+    // discontinuous Galerkin doesn't require an artificial viscosity
     if (advection_field.is_discontinuous(introspection))
       return 0.;
 
+    std::vector<double> residual = assemblers->compute_advection_system_residual(*scratch.explicit_material_model_inputs.cell,
+                                                                                 advection_field,
+                                                                                 scratch);
+
     double max_residual = 0;
     double max_velocity = 0;
-    double max_density = 0;
-    double max_specific_heat = 0;
+    double max_density = (advection_field.is_temperature()) ? 0.0 : 1.0;
+    double max_specific_heat = (advection_field.is_temperature()) ? 0.0 : 1.0;
     double max_conductivity = 0;
 
-    compute_advection_system_residual(scratch,
-                                      average_field,
-                                      advection_field,
-                                      max_residual,
-                                      max_velocity,
-                                      max_density,
-                                      max_specific_heat,
-                                      max_conductivity);
+    for (unsigned int q=0; q < scratch.old_velocity_values.size(); ++q)
```

Comment:
To make the code easier to read, could you query the loop upper bound via something that clearly indicates that it is the number of quadrature points? It should be easy to get this out of the fevalues object in scratch, for example